mllib/src/main/scala/org/apache/spark/mllib/fpm/AssociationRules.scala

```diff
@@ -88,6 +110,204 @@ class AssociationRules private[fpm] (
+
+    val initCandidates = freqItemIndices.flatMap {
+      case (indices, freq) =>
+        indices.flatMap {
+          index =>
+            indices.partition(_ == index) match {
+              case (consequent, antecendent) if antecendent.nonEmpty =>
+                Some((antecendent, (consequent, freq)))
+              case _ => None
+            }
+        }
+    }
+
+    val initRules = sc.emptyRDD[(Seq[Int], Seq[Int], Long, Long)]
+
+    @tailrec
+    def loop(candidates: RDD[(Seq[Int], (Seq[Int], Long))],
+             lenConsequent: Int,
+             rules: RDD[(Seq[Int], Seq[Int], Long, Long)]
+            ): RDD[(Seq[Int], Seq[Int], Long, Long)] = {
+
+      val numCandidates = candidates.count()
+
+      log.info(s"Candidates for ${lenConsequent}-consequent rules : ${numCandidates}")
+
+      if (numCandidates == 0 || lenConsequent > maxConsequent) {
+        rules
+      } else {
+        val newRules = candidates.join(freqItemIndices).filter{
+          case (antecendent, ((consequent, freqUnion), freqAntecedent)) =>
+            freqUnion >= minConfidence * freqAntecedent
+        }.map {
+          case (antecendent, ((consequent, freqUnion), freqAntecedent)) =>
+            (antecendent, consequent, freqUnion, freqAntecedent)
+        }
+
+        val numNewRules = newRules.count()
+        log.info(s"Generated ${lenConsequent}-consequent rules : ${numNewRules}")
+
+        if (numNewRules == 0) {
+          rules
+        } else if (lenConsequent == maxConsequent) {
+          sc.union(rules, newRules)
+        } else {
+          val numPairs = (lenConsequent + 1) * lenConsequent / 2
+          val newCandidates = newRules.filter{
+            // rules whose antecendent's length equals to 1 can not be used to generate new rules
+            case (antecendent, consequent, freqUnion, freqAntecedent) =>
+              antecendent.size > 1
+          }.map {
+            case (antecendent, consequent, freqUnion, freqAntecedent) =>
+              val union = seqAdd(antecendent, consequent)
+              ((union, freqUnion), consequent)
+          }.groupByKey().filter {
+            // Rule pruning. For a (lenConsequent + 1)-consequent rules, at least
+            // (lenConsequent + 1) lenConsequent-consequent rules are needed.
+            case ((union, freqUnion), consequents) =>
+              consequents.size > lenConsequent
+          }.flatMap {
+            case ((union, freqUnion), consequents) =>
+              val array = consequents.toArray
+              val newConsequentCount = collection.mutable.Map[Seq[Int], Int]()
+              for (i <- 0 until array.length; j <- i + 1 until array.length) {
+                val newConsequent = seqAdd(array(i), array(j))
+                if (newConsequent.length == lenConsequent + 1) {
+                  val cnt = newConsequentCount.getOrElse(newConsequent, 0)
+                  newConsequentCount.update(newConsequent, cnt + 1)
+                }
+              }
```

Comment:
This paragraph is too long. We should break it down and give meaningful names to the intermediate result.